File: ui/name_sequence.py
from __future__ import annotations

import re
import unicodedata


def make_sequence(base_name: str, count: int) -> list[str]:
    name = base_name.strip()
    count = max(1, count)
    if count == 1:
        return [name]

    match = re.search(r"(\d+)(?!.*\d)", name)
    if match is None:
        return [f"{name}-{index}" for index in range(1, count + 1)]

    start = int(match.group(1))
    width = len(match.group(1))
    prefix = name[: match.start(1)]
    suffix = name[match.end(1) :]
    return [f"{prefix}{str(start + offset).zfill(width)}{suffix}" for offset in range(count)]


def next_name_after_sequence(base_name: str, count: int) -> str:
    names = make_sequence(base_name, count + 1)
    return names[-1]


def next_name_for_prefix(prefix: str, existing_names: list[str]) -> str:
    rule = prefix.strip() or "ITEM-N"
    pattern = name_rule_pattern(rule)
    next_number = next_number_for_pattern(pattern, existing_names)
    return f"{pattern.prefix}{str(next_number).zfill(pattern.width)}{pattern.suffix}"


def next_number_for_prefix(prefix: str, existing_names: list[str]) -> int:
    escaped = re.escape(prefix.strip())
    pattern = re.compile(rf"^{escaped}[-_ ]?(\d+)$", re.IGNORECASE)
    highest = 0
    for name in existing_names:
        match = pattern.match(name.strip())
        if match:
            highest = max(highest, int(match.group(1)))
    return highest + 1


class NameRulePattern:
    def __init__(self, prefix: str, suffix: str, width: int) -> None:
        self.prefix = prefix
        self.suffix = suffix
        self.width = width


def name_rule_pattern(rule: str) -> NameRulePattern:
    match = re.search(r"N+", rule, re.IGNORECASE)
    if match:
        return NameRulePattern(rule[: match.start()], rule[match.end() :], match.end() - match.start())
    return NameRulePattern(f"{rule}-", "", 1)


def next_number_for_pattern(pattern: NameRulePattern, existing_names: list[str]) -> int:
    escaped_prefix = re.escape(pattern.prefix)
    escaped_suffix = re.escape(pattern.suffix)
    regex = re.compile(rf"^{escaped_prefix}(\d+){escaped_suffix}$", re.IGNORECASE)
    highest = 0
    for name in existing_names:
        match = regex.match(name.strip())
        if match:
            highest = max(highest, int(match.group(1)))
    return highest + 1


def resolve_name_rule(rules: dict[str, str], item_type: str, fallback: str) -> str:
    if item_type in rules:
        return rules[item_type]
    normalized_type = normalize_key(item_type)
    for key, value in rules.items():
        if normalize_key(key) == normalized_type:
            return value
    return fallback


def normalize_key(value: str) -> str:
    text = unicodedata.normalize("NFKD", value)
    text = "".join(char for char in text if not unicodedata.combining(char))
    return re.sub(r"[^a-z0-9]+", "", text.lower())

File: flutter/lib/domain/services/name_sequence.dart
final _lastNumber = RegExp(r'(\d+)(?!.*\d)');
final _rulePlaceholder = RegExp('N+', caseSensitive: false);

List<String> makeSequence(String baseName, int count) {
  final name = baseName.trim();
  final safeCount = count < 1 ? 1 : count;
  if (safeCount == 1) {
    return [name];
  }

  final match = _lastNumber.firstMatch(name);
  if (match == null) {
    return [
      for (var index = 1; index <= safeCount; index += 1) '$name-$index',
    ];
  }

  final start = int.parse(match.group(1)!);
  final width = match.group(1)!.length;
  final prefix = name.substring(0, match.start);
  final suffix = name.substring(match.end);
  return [
    for (var offset = 0; offset < safeCount; offset += 1)
      '$prefix${(start + offset).toString().padLeft(width, '0')}$suffix',
  ];
}

String nextNameForRule(
  Map<String, String> rules,
  String itemType,
  String fallback,
  List<String> existingNames,
) {
  final rule = resolveNameRule(rules, itemType, fallback);
  final pattern = nameRulePattern(rule);
  final nextNumber = nextNumberForPattern(pattern, existingNames);
  return '${pattern.prefix}${nextNumber.toString().padLeft(pattern.width, '0')}${pattern.suffix}';
}

NameRulePattern nameRulePattern(String rule) {
  final match = _rulePlaceholder.firstMatch(rule);
  if (match == null) {
    return NameRulePattern('$rule-', '', 1);
  }
  return NameRulePattern(
    rule.substring(0, match.start),
    rule.substring(match.end),
    match.end - match.start,
  );
}

int nextNumberForPattern(NameRulePattern pattern, List<String> existingNames) {
  final regex = RegExp(
    '^${RegExp.escape(pattern.prefix)}(\\d+)${RegExp.escape(pattern.suffix)}\$',
    caseSensitive: false,
  );
  var highest = 0;
  for (final name in existingNames) {
    final match = regex.firstMatch(name.trim());
    if (match != null) {
      final number = int.tryParse(match.group(1)!);
      if (number != null && number > highest) {
        highest = number;
      }
    }
  }
  return highest + 1;
}

String resolveNameRule(
  Map<String, String> rules,
  String itemType,
  String fallback,
) {
  if (rules.containsKey(itemType)) {
    return rules[itemType]!;
  }
  final normalizedType = normalizeKey(itemType);
  for (final entry in rules.entries) {
    if (normalizeKey(entry.key) == normalizedType) {
      return entry.value;
    }
  }
  return fallback;
}

String normalizeKey(String value) {
  final mapped = value.toLowerCase().split('').map(_withoutPolishMark).join();
  return mapped.replaceAll(RegExp(r'[^a-z0-9]+'), '');
}

String _withoutPolishMark(String value) {
  return switch (value) {
    'ą' => 'a',
    'ć' => 'c',
    'ę' => 'e',
    'ł' => 'l',
    'ń' => 'n',
    'ó' => 'o',
    'ś' => 's',
    'ż' => 'z',
    'ź' => 'z',
    _ => value,
  };
}

class NameRulePattern {
  const NameRulePattern(this.prefix, this.suffix, this.width);

  final String prefix;
  final String suffix;
  final int width;
}
